### ISSUE TITLE
GroupMemoryBarrier? GroupMemoryBarrierWithGroupSync?

### ISSUE THREAD

**Issue by pengliu916** [Microsoft/DirectX-Graphics-Samples#140]
I have seen miniEngine use XXXMemoryBarrierWithGroupSync exclusively across all its cs where there need a 'sync'. But after looking through several of them I found most of them just need memory barrier, so it may be benificial (conceptually) to use XXXMemoryBarrier instead of XXXMemoryBarrierWithGroupSync.  Or are there some reasons to stick with the GroupSync version which I missed? Or am I misunderstand XXXMemoryBarrierWithGroupSync?

Thanks
Peng

**Comment by stanard** [comment]
Great observation!  I'm responsible for almost all of those group syncs, and I tend to default to the GroupMemoryBarrierWithGroupSync() call.  You're probably aware of the difference of the various barriers, but I'll briefly paraphrase the differences.

GroupMemoryBarrier - Wait for all issued reads and writes from groupshared memory (LDS) to finish
DeviceMemoryBarrier - Same but for system/video memory
AllMemoryBarrier - Flush any and all memory transactions (Group + Device)

Now appending "GroupSync" to the call also synchronizes all of the threads so that they've executed all of the same instructions.  This can result in subtle race conditions if left off.  Suppose you do a GroupMemoryBarrier _without_ a GroupSync.  If not all of the threads are at the same program counter (PC), then some of those threads might not have issued their read or write instructions before saying "wait for all that have been issued".  To be safe and make sure you wait for all memory transactions that you **expect** to be issued, you need to get all of your threads in sync.

That concern is really only valid when your group thread count is larger than the GPU's native wave size.  All of the threads in the same wave are naturally at the same instruction.  It's only when thread groups are broken up into multiple waves that you have threads issuing different instructions.  But in the case where your thread count equals the wave size, the shader compiler can silently optimize out your group sync because it knows its not necessary.

So long story short, it's better to be safe.  And it's hard to be too safe in this case because the compiler will ignore your request when it's not necessary.

I wanted to point out one more thing.  I almost never use a Device or AllMemoryBarrier because I tend to read from system memory and write to groupshared memory.  By the time the write has happened, the read from device memory must also have happened.  If I was just reading into registers, I would need a DeviceMemoryBarrier.

Cheers,
James

**Comment by pengliu916** [comment]
Thanks James for such detailed explanation, that's really informative. But what would be the proper and safe case to use GroupMemoryBarrier then, since compiler can opt out the GroupSync when threads are within one wavefront/warp. Or GroupMemoryBarrier will be deprecated later?

**Comment by stanard** [comment]
You might need inter-thread synchronization only within a wave or smaller.  For example, maybe your thread group size is 256, but your threads are operating on data in groups of 4x4.  You need to flush writes before reads or vice versa, but you can assume all threads working on the same set of data are doing so in lock-step.  So MemoryBarrior is useful but not GroupSync.

**Comment by pengliu916** [comment]
ah, yup. Thanks for all your time answering my questions :-)

**Comment by nikitablack** [comment]
Sorry for asking in a closed issue but that's exactly the problem I'm trying to understand for a couple of weeks. I tried to compare gpu memory barriers with c++11 memory fences with different memory ordering (relaxed, seq_consistency, acquire/release) but I can't see the similarities. @stanard, after your great explanation memory barriers start to make some sense to me thought not absolutely clear. Can you please explain it one more with an example. Let's take this pseudocode:
```
shaderResource.store(42); // #1
GroupMemoryBarrier(); // #2
int i = shaderResource.load(); //#3
```
Am I correctly understanding that if `thread1` and `thread2` (a hypothetical group made of 2 threads) are from different waves than this flow is possible?

1. `thread1` writes to the resource (`#1`). Since `thread2` is from another wave it doesn't go with `thread1` in a lockstep and it doing nothing for now (I mean it's possible to do nothing).
2. `thread1` reaches barrier (`#2`). No more threads in a group accessed memory untill this moment - only `thread1`. That means that this barrier can be passed for this thread. `thread2` still doing nothing.
3. `thread1` reads from the memory (`#3`). Now, at the same time, `thread2` starts to write (`#1`). Ans it's a classical race condition.

Is my understanding correct?

**Comment by stanard** [comment]
Hi @nikitablack,

I think your understanding is there, but you need a DeviceMemoryBarrier() when writing to memory.

You are correct that because thread2 comes from a different wave, it could write to the same location that thread1 is about to reload.  You need a GroupSync to guarantee that both threads have issued their write instructions.  But it's generally a bad thing to have multiple threads writing to the same location anyway.

I should also point out that results could change if your threads came from different thread groups because then they could be running on different compute units, execute at different times, and even go through different L1 caches.  And there are no dispatch-wide memory barriers, so you couldn't synchronize those two threads.

When it comes to multi-threaded GPU memory accesses, try to have threads be responsible for their own data (like SIMD).  Use groupshared memory to share information between threads for things like scatter, gather, convolutions, etc.  If you want to write a single value as the result of your dispatch, put a branch around it:  if (DispatchThreadID == 0) { // write value }

Good luck!  Welcome to mind-bending world of very parallel algorithms.  :-)

-James

**Comment by nikitablack** [comment]
Thank you @stanard, I think I got it.
Just to summarize (taking the example from my previous post):
- if we have two threads from the same group and the same wave - there're no chances of data race. We don't need to sync here.
- if we have two threads from the same group but from different waves - there's a chance of data race. We need to sync here.
- if we have two threads from different groups - we can't share the same group memory (it's obvious), but we can share device memory. But there're no such barriers in d3d - all barriers work only in the same group. We need another sync mechanism, but not barriers. Atomic counters maybe.

And yes, this all applies only to situation where we have a write to the **same** data (as in my example). As @stanard pointed out:
> it's generally a bad thing to have multiple threads writing to the same location anyway.
